In [1]:
import pandas as pd

# Load the data
file_path = r"C:\Users\Tejus\Downloads\arpit\Sample - Superstore.csv"
superstore_data = pd.read_csv(file_path)

# Display the first few rows of the dataframe to understand its structure
superstore_data.head()


UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 2944: invalid start byte

In [2]:
# Attempting to read the file with a different encoding
try:
    superstore_data = pd.read_csv(file_path, encoding='ISO-8859-1')
except Exception as e:
    error_message = str(e)

# Check if the file was loaded successfully or return the error message
if 'superstore_data' in locals():
    loaded_successfully = True
    superstore_data.head()
else:
    loaded_successfully = False
    error_message



In [3]:
# Displaying the basic information about the dataframe
superstore_data.info()

# Displaying statistical summary of numerical columns
superstore_data.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

Row ID   Postal Code         Sales     Quantity     Discount  \
count  9994.000000   9994.000000   9994.000000  9994.000000  9994.000000   
mean   4997.500000  55190.379428    229.858001     3.789574     0.156203   
std    2885.163629  32063.693350    623.245101     2.225110     0.206452   
min       1.000000   1040.000000      0.444000     1.000000     0.000000   
25%    2499.250000  23223.000000     17.280000     2.000000     0.000000   
50%    4997.500000  56430.500000     54.490000     3.000000     0.200000   
75%    7495.750000  90008.000000    209.940000     5.000000     0.200000   
max    9994.000000  99301.000000  22638.480000    14.000000     0.800000   

            Profit  
count  9994.000000  
mean     28.656896  
std     234.260108  
min   -6599.978000  
25%       1.728750  
50%       8.666500  
75%      29.364000  
max    8399.976000

In [4]:
# Grouping data by Product ID and Product Name to aggregate their total sales and profit
product_profitability = superstore_data.groupby(['Product ID', 'Product Name', 'Category', 'Sub-Category']).agg(
    Total_Sales=pd.NamedAgg(column='Sales', aggfunc='sum'),
    Total_Profit=pd.NamedAgg(column='Profit', aggfunc='sum')
).reset_index()

# Sorting products by Total Profit to identify the lowest profit or loss-making products
low_profit_products = product_profitability.sort_values(by='Total_Profit').head(10)
low_profit_products


Product ID                                       Product Name  \
1645  TEC-MA-10000418          Cubify CubeX 3D Printer Double Head Print   
1650  TEC-MA-10000822          Lexmark MX611dhe Monochrome Laser Printer   
1696  TEC-MA-10004125          Cubify CubeX 3D Printer Triple Head Print   
326   FUR-TA-10000198  Chromcraft Bull-Nose Wood Oval Conference Tabl...   
344   FUR-TA-10001889  Bush Advantage Collection Racetrack Conference...   
858   OFF-BI-10004995          GBC DocuBind P400 Electric Binding System   
1669  TEC-MA-10002412  Cisco TelePresence System EX90 Videoconferenci...   
1465  OFF-SU-10002881  Martin Yale Chadless Opener Electric Letter Op...   
346   FUR-TA-10001950                       Balt Solid Wood Round Tables   
375   FUR-TA-10004289  BoxOffice By Design Rectangular and Half-Moon ...   

             Category Sub-Category  Total_Sales  Total_Profit  
1645       Technology     Machines    11099.963    -8879.9704  
1650       Technology     Machines    16829.901    -4589.9730  
1696       Technology     Machines     7999.980    -3839.9904  
326         Furniture       Tables     9917.640    -2876.1156  
344         Furniture       Tables     9544.725    -1934.3976  
858   Office Supplies      Binders    17965.068    -1878.1662  
1669       Technology     Machines    22638.480    -1811.0784  
1465  Office Supplies     Supplies    16656.200    -1299.1836  
346         Furniture       Tables     6518.754    -1201.0581  
375         Furniture       Tables     1706.250    -1148.4375

In [7]:
# Grouping data by Product ID and Product Name to aggregate their total quantity sold and total profit
product_performance = superstore_data.groupby(['Product ID', 'Product Name', 'Category', 'Sub-Category']).agg(
    Total_Quantity=pd.NamedAgg(column='Quantity', aggfunc='sum'),
    Total_Sales=pd.NamedAgg(column='Sales', aggfunc='sum'),
    Total_Profit=pd.NamedAgg(column='Profit', aggfunc='sum')
).reset_index()

# Identifying products with low sales numbers and low profitability
# Filtering criteria can be adjusted as needed
low_sales_and_profit = product_performance[(product_performance['Total_Quantity'] < product_performance['Total_Quantity'].quantile(0.25)) & 
                                           (product_performance['Total_Profit'] < product_performance['Total_Profit'].quantile(0.25))]

low_sales_and_profit.sort_values(by=['Total_Quantity', 'Total_Profit']).head(10)


Product ID                                       Product Name  \
17    FUR-BO-10002206  Bush Saratoga Collection 5-Shelf Bookcase, Han...   
1688  TEC-MA-10003493                Penpower WorldCard Pro Card Scanner   
1029  OFF-PA-10000048                                           Xerox 20   
568   OFF-AR-10002704              Boston 1900 Electric Pencil Sharpener   
1663  TEC-MA-10001856                              Okidata C610n Printer   
1671  TEC-MA-10002694  Hewlett-Packard Deskjet F4180 All-in-One Color...   
427   OFF-AP-10002203  Eureka Disposable Bags for Sanitaire Vibra Gro...   
1885  TEC-PH-10004875                  PNY Rapid USB Car Charger - Black   
875   OFF-EN-10001535                                Grip Seal Envelopes   
1260  OFF-PA-10003919                                         Xerox 1989   

             Category Sub-Category  Total_Quantity  Total_Sales  Total_Profit  
17          Furniture    Bookcases               1      119.833      -12.6882  
1688       Technology     Machines               1       91.475       -1.8295  
1029  Office Supplies        Paper               1        6.480        3.1104  
568   Office Supplies          Art               1       14.980        4.4940  
1663       Technology     Machines               2      649.000     -272.5800  
1671       Technology     Machines               2      101.994      -71.3958  
427   Office Supplies   Appliances               2        1.624       -4.4660  
1885       Technology       Phones               2        9.588       -2.0774  
875   Office Supplies    Envelopes               2        7.072        2.3868  
1260  Office Supplies        Paper               2        7.968        2.6892

Here are some details about these products:

Bush Saratoga Collection 5-Shelf Bookcase, Hanover Cherry (Furniture, Bookcases): Sold only 1 unit with a loss of $12.69.
Penpower WorldCard Pro Card Scanner (Technology, Machines): Sold only 1 unit with a minor loss.
Xerox 20 (Office Supplies, Paper): Sold only 1 unit but with a small profit.
Boston 1900 Electric Pencil Sharpener (Office Supplies, Art): Sold only 1 unit with a small profit.
Okidata C610n Printer (Technology, Machines): Sold 2 units with a significant loss.
Hewlett-Packard Deskjet F4180 All-in-One Color Printer (Technology, Machines): Sold 2 units with a considerable loss.
Eureka Disposable Bags for Sanitaire Vibra Groomer I (Office Supplies, Appliances): Sold 2 units with a loss.
PNY Rapid USB Car Charger - Black (Technology, Phones): Sold 2 units with a minor loss.
Grip Seal Envelopes (Office Supplies, Envelopes): Sold 2 units with a small profit.
Xerox 1989 (Office Supplies, Paper): Sold 2 units with a small profit.